src/Client.php

```diff
@@ -0,0 +1,370 @@
+      return $endpoint->list();
+    }
+
+    if (is_int($arg)) {
+      return $endpoint->retrieve($arg);
+    }
+
+    if (is_array($arg)) {
+      if (! $endpoint instanceof ReadWritable) {
+        throw new ApiException(
+          ApiException::ENDPOINT_NOT_WRITABLE,
+          ['endpoint' => $name]
+        );
+      }
+
+      return $endpoint->create($arg);
+    }
+
+    if ($arg instanceof Model) {
+      if (! $endpoint instanceof ReadWritable) {
+        throw new ApiException(
+          ApiException::ENDPOINT_NOT_WRITABLE,
+          ['endpoint' => $name]
+        );
+      }
+
+      return $endpoint->update($arg);
+    }
+
+    throw new SdkException(
+      SdkException::WRONG_CALL_ARG,
+      [
+        'class' => __CLASS__,
+        'expected' => 'null|int|array|Model',
+        'type' => Util::type($arg)
+      ]
+    );
+  }
+
+  /**
+   * @see https://php.net/__get
+   * Allows endpoints to be accessed as properties.
+   *
+   * @example <?php
+   *  // get the "api-token" endpoint
+   *  $endpoint = $Client->ApiToken;
+   *
+   * @param string $name Endpoint classname (short name or fully qualified)
+   * @return Endpoint on success
+   * @throws SdkException If Endpoint is unknown
+   */
+  public function __get(string $name) : Endpoint {
+    return $this->getEndpoint($name);
+  }
+
+  /**
+   * Gets an API endpoint.
+   *
+   * For convenience, endpoints can also be accessed as Client properties.
+   * @see Client::__get
+   *
+   * @param string $name Endpoint classname (short name or fully qualified)
+   * @return Endpoint
+   * @throws SdkException If the endpoint is unknown
+   */
+  public function getEndpoint(string $name) : Endpoint {
+    if (empty($this->_endpoints[$name])) {
+      $this->_endpoints[$name] = $this->_newEndpoint($name);
+    }
+
+    return $this->_endpoints[$name];
+  }
+
+  /**
+   * Perform an API request.
+   *
+   * This is intended for use by Endpoints,
+   * and should generally not be used otherwise.
+   *
+   * Requests and responses are logged
+   * if the "debug" or "request.log" config options are true.
+   *
+   * @param string $method The http method to use
+   * @param string $endpoint The API endpoint to request
+   * @param array $params Http client parameters
+   * @return Response
+   * @throws ApiException On http error (4xx, 5xx, network issues, etc.)
+   * @throws SdkException On any other error
+   */
+  public function request(
+    string $method,
+    string $endpoint,
+    array $params = []
+  ) : Response {
+    try {
+      $params['headers'] =
+        ($params['headers'] ?? []) + $this->_getDefaultHeaders();
+
+      $response = new Response(
+        $this->_client->request($method, $endpoint, $params)
+      );
+      return $response;
+
+    } catch (ConnectException $e) {
+      throw new ApiException(ApiException::CANNOT_CONNECT, $e);
+    } catch (ClientException $e) {
+      switch ($e->getResponse()->getStatusCode()) {
+        case 401 :
```

Comment:
extra space after `401`

Reply:
yep.  i'm already working on all the cs stuff travis found.